Io abstraction

P-Storm opened the pull request:
## Summary of the Pull Request

This makes that the FileSystem mockable.

## PR Checklist
* [X] Applies to  #6920
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [X] Tests added/passed

## Info on Pull Request

This creates creates a possability to test with a mockable filesystem. This helps testing the Folder Plugins, saving/loading settings, but also for other areas.

* I did split `Query_WhenCalled` in recursive and non recursive.
* `ISettingsPath` path manipulations was easier to create an interface for it
* Removed `GetMockIOProviderForSaveLoadExists` because MockFileSystem is taking care of this.

## Validation Steps Performed

Do the unit tests

Commit:
Initial IO Abstractions

Commit:
Merge branch 'master' into IO-Abstraction

# Conflicts:
#	src/core/Microsoft.PowerToys.Settings.UI.UnitTests/Microsoft.PowerToys.Settings.UI.UnitTests.csproj
#	src/core/Microsoft.PowerToys.Settings.UI.UnitTests/ViewModelTests/FancyZones.cs
#	src/core/Microsoft.PowerToys.Settings.UI.UnitTests/ViewModelTests/General.cs
#	src/core/Microsoft.PowerToys.Settings.UI.UnitTests/ViewModelTests/ImageResizer.cs
#	src/modules/launcher/Plugins/Microsoft.Plugin.Folder/Sources/QueryFileSystemInfo.cs
#	src/modules/launcher/Wox.Infrastructure/FileSystemHelper/FileWrapper.cs

Commit:
IO.Abstraction changed

Commit:
Merge branch 'master' into IO-Abstraction

# Conflicts:
#	src/core/Microsoft.PowerToys.Settings.UI.UnitTests/ViewModelTests/ImageResizer.cs
#	src/core/Microsoft.PowerToys.Settings.UI/Views/PowerPreviewPage.xaml.cs

Commit:
Fixed Inconclusive and MockFileSystems

Commit:
Merge branch 'master' into IO-Abstraction

# Conflicts:
#	src/modules/launcher/Plugins/Microsoft.Plugin.Folder/Sources/Result/FileItemResult.cs

Commit:
And merge resolve

P-Storm commented:
@crutkas I can't request a code review of this, is there a way to enable it?

crutkas commented:
/azp run

ryanbodrug-microsoft commented:
> @crutkas I can't request a code review of this, is there a way to enable it? I have merged it again to master.

I'm taking a look now.  Sorry for the delay @crutkas actually asked me to look at this last week.

enricogior commented:
Should we split this in smaller PRs so that they can be reviewed and tested more easily?

Commit:
Merge branch 'master' into IO-Abstraction

# Conflicts:
#	src/core/Microsoft.PowerToys.Settings.UI.Lib/ViewModels/GeneralViewModel.cs

P-Storm commented:
> Should we split this in smaller PRs so that they can be reviewed and tested more easily?
It is probably hard to split the PR right now. If people want it I would have to cherry pick it into multiple PR's, but I have also never done that for a PR that is already in review.

And remerged the merge conflict. This time it was not that bad of a merge conflict.

ryanbodrug-microsoft commented:
> > @crutkas I can't request a code review of this, is there a way to enable it? I have merged it again to master.
> 
> I'm taking a look now. Sorry for the delay @crutkas actually asked me to look at this last week.

@P-Storm would you be able to fix the compile errors so I can test locally?

P-Storm commented:
> > > @crutkas I can't request a code review of this, is there a way to enable it? I have merged it again to master.
> > 
> > 
> > I'm taking a look now. Sorry for the delay @crutkas actually asked me to look at this last week.
> 
> @P-Storm would you be able to fix the compile errors so I can test locally?

Dammit I did fix this, will fix this merge issue again :)

Commit:
Fixed error in 'GeneralPage.xaml.cs' and Merge branch 'master' into IO-Abstraction

# Conflicts:
#	src/modules/launcher/Wox.Core/Plugin/PluginInstaller.cs
#	src/modules/launcher/Wox.Core/Plugin/PluginManager.cs

Commit:
Alphabethically ordering of namespace SA1210 (& removed not used namespace.

P-Storm commented:
@ryanbodrug-microsoft you can test it now :) . Hopefull it won't be blocked by a new Merge Conflict.

ryanbodrug-microsoft commented:
> @ryanbodrug-microsoft you can test it now :) . Hopefully it won't be blocked by a new Merge Conflict.

Thanks @P-Storm looking now.  Sorry I know this is a pain, but much appreciated 🥇

Commit:
Merge branch 'master' into IO-Abstraction

# Conflicts:
#	src/modules/launcher/Plugins/Microsoft.Plugin.Program/Storage/Win32ProgramRepository.cs

P-Storm commented:
> 
> 
> > @ryanbodrug-microsoft you can test it now :) . Hopefully it won't be blocked by a new Merge Conflict.
> 
> Thanks @P-Storm looking now. Sorry I know this is a pain, but much appreciated 🥇

And again. This time git was stupid usings problems. Let's try it again :)

ryanbodrug-microsoft reviewed (commented):
It looks good to me.   A couple really small things. 

I've verified KBM mapping keys/shortcuts, as well as dynamically adding and removing win32 and packaged apps, as well as updating settings files.   I'll defer to @enricogior  for final approval as his team will be on the hook for maintaining these files after this month.

Commit:
Review resolution: Removed Reference Of System.IO.Abstractions (with HintPath)

P-Storm commented on src/modules/previewpane/common/PreviewHandlerCommon.csproj:
```diff
@@ -86,6 +86,9 @@
     <Reference Include="System.Drawing" />
+    <Reference Include="System.IO.Abstractions">
+      <HintPath>C:\Users\P-Storm\.nuget\packages\system.io.abstractions\12.2.1\lib\net461\System.IO.Abstractions.dll</HintPath>
+    </Reference>
```
Solved in latest commit

P-Storm commented on src/modules/previewpane/MarkdownPreviewHandler/MarkdownPreviewHandler.csproj:
```diff
@@ -86,6 +86,9 @@
     <Reference Include="System.Drawing" />
+    <Reference Include="System.IO.Abstractions">
+      <HintPath>C:\Users\P-Storm\.nuget\packages\system.io.abstractions\12.2.1\lib\net461\System.IO.Abstractions.dll</HintPath>
+    </Reference>
```
Solved in latest commit

P-Storm commented:
> 
> 
> It looks good to me. A couple really small things.
> 
> I've verified KBM mapping keys/shortcuts, as well as dynamically adding and removing win32 and packaged apps, as well as updating settings files. I'll defer to @enricogior for final approval as his team will be on the hook for maintaining these files after this month.

And did fixed those small things. This should be completely ready. :)

crutkas commented:
We want to 'lock' 0.25 this Friday, due to size, we feel we should get this in 0.27 as soon as we snap to stable.  This gives everyone a month to verify all the code path updates here.

P-Storm commented:
Ugh more merge requests. Can someone look again to my pull request and merge when I have it mergable again? I think 0.25 is out so this could be closed than :)

crutkas commented:
we can do this now.

ryanbodrug-microsoft commented:
@enricogior Let me know if you want me to approve this.  This may conflict with some work on InVEST so I would like to get this in as soon as possible, but I don't want to make the call on behalf of your team.

enricogior commented:
Let's get this PR updated to resolve the conflicts and then we can do the final review and merge.

Commit:
Merge fixed again.

P-Storm commented:
@enricogior I have fixed the merge request again. 

I see that there could be some work in tests that are using IIOProviderMocks, but that will be something for another PR.

(IIOProviderMocks is not using the Mocked file system, but it uses the interfaces. This should be also using the mocked filesystem).

enricogior commented:
@P-Storm 
thanks!

There is one very important thing that we forgot to verify: do you know if System.IO.Abstractions is fully supported/tested on ARM?

P-Storm commented:
> 
> 
> @P-Storm
> thanks!
> 
> There is one very important thing that we forgot to verify: do you know if System.IO.Abstractions is fully supported/tested on ARM?

That is a great question, that I do not know the answer or know how it could be tested. I know that it is a wrapper around the normal System.IO.

crutkas commented:
Validating but pretty sure since it is .NET standard 2.1 this should be ok since this library would be functional with a UWP app which by default supports ARM64

enricogior commented:
@P-Storm 
today we do the final review of your PR and then merge it.
As you can see, there new conflicts, let's plan a final conflict resolution and then merge it right after they are resolved to avoid more conflicts.

yuyoyuppe reviewed (approved):
I've verified the branch locally on the case of FZ: settings are editable and are saved correctly. The whole solution also compiled correctly.
Thank for the contribution! As @enricog mentioned, we still need to resolve the merge conflicts before merging this.

P-Storm commented on src/core/Microsoft.PowerToys.Settings.UI.Library/ISettingsPath.cs:
```diff
@@ -0,0 +1,17 @@
+
+        void CreateSettingsFolder(string powertoy);
+
+        void DeleteSettings(string powertoy = "");
```
This was already the case when I introduced the abstraction, I didn't change a thing in this. So I guess it is possible, but didn't want to change this in my PR.

P-Storm commented on src/core/Microsoft.PowerToys.Settings.UI.Library/SettingPath.cs:
```diff
@@ -0,0 +1,62 @@
+
+        public bool SettingsFolderExists(string powertoy)
+        {
+            return _directory.Exists(System.IO.Path.Combine(LocalApplicationDataFolder(), $"Microsoft\\PowerToys\\{powertoy}"));
```
That is fine by me, but I didn't introduce this change, so I didn't want to include this in my PR.

P-Storm commented:
For the review, thanks @yuyoyuppe for looking into it. As explained I didn't introduce the changes so I'm hesistant to introduce those in my PR. If we want to change this (and maybe other things in these areas, I would suggest a loose PR for it.

For the conflicting files, I will do this right after my work (in about 1 hour), and push it again. Will mention you again @enricogior  than, so we can merge it fast to master and not have another merge request :)

yuyoyuppe commented:
@P-Storm sure, since those changes weren't introduced by you, I've resolved the comments for now, we'll do them as a separate PRs later.

Commit:
Merge branch 'master' into IO-Abstraction

# Conflicts:
#	src/core/Microsoft.PowerToys.Settings.UI.UnitTests/BackwardsCompatibility/BackCompatTestProperties.cs
#	src/core/Microsoft.PowerToys.Settings.UI.UnitTests/Mocks/IIOProviderMocks.cs
#	src/core/Microsoft.PowerToys.Settings.UI/Views/KeyboardManagerPage.xaml.cs
#	src/modules/colorPicker/ColorPickerUI/Helpers/Logger.cs
#	src/modules/launcher/Plugins/Microsoft.Plugin.Folder.UnitTests/InternalQueryFolderTests.cs
#	src/modules/launcher/Plugins/Microsoft.Plugin.Folder/Sources/Path/DriveInformation.cs
#	src/modules/launcher/PowerLauncher/SettingsWatcher.cs
#	src/modules/launcher/Wox.Infrastructure/Image/ImageLoader.cs
#	src/modules/launcher/Wox.Infrastructure/Storage/BinaryStorage`1.cs
#	src/modules/launcher/Wox.Plugin/SharedCommands/FilesFolders.cs

P-Storm commented:
@enricogior I have merged it!

enricogior commented:
@P-Storm 
great!
Waiting for CI to complete and I'll merge it.
